proposals/0000-ghc-extensions.rst

```diff
@@ -0,0 +1,120 @@
+
+
+Proposed Change Specification
+-----------------------------
+
+We propose to introduce a process to develop new "language versions", dubbed ``GHC20XX``, which extend ``Haskell2010`` with some default extensions. This proposal does not set *which* extensions are included in the first version, but rather seeks to provide a framework to reach a conclusion in a constructive and timely manner.
+
+Criteria
+^^^^^^^^
+
+We define a set of criteria that an extension should guide the committee members in their process of picking suitable extensions. These are not hard criteria, but often subjective, and not all need to be satisfied in all cases.
+
+1. The extension is (mostly) *conservative*: All programs that were accepted without the extension remain accepted, and with the same meaning. We prefer to take this as a strict criteria, and grant exceptions on a case-by-case basis.
+2. *New failure modes* that become possible with the extension are *rare* and/or easy to diagnose. These failure modes include new error messages, wrong inferred types, and runtime errors, for example. In other words, the *developer experience* should not get worse by adding this extension to the default mix.
+3. The extensions *complement* the design of standard Haskell. For example, ``MultiParamTypeClasses`` extends what's already there, and works well with the rest of features. If we suddenly decided you can use Lisp-like syntax, that would not complement the current design.
+4. The extension has been – and can reasonably be predicted to remain – *stable*.
+5. The extension is one that users might plausibly want to be on all the time. This excludes experimental extensions that deliberately enable potentially-unsafe or unstable features, such as ``IncoherentInstances`` or ``MagicHash``.
+6. The extension has *widespread* usage.
+7. The extension is favored by the community, with many in favor, and very few opposed to its inclusion.
+
+
+Process
+^^^^^^^
+
+* 4 months before the expected GHC spring release day of 202x, the committee Secretary starts the GHC202x process.
+
+  They inform the committee, in an email to the mailing list, of all language extensions supported by the latest released GHC that are not in GHC202(x-1), which could be added. They also list all extensions *in* GHC202(x-1), which might be omitted in GHC202(x-1) (likely a rare thing).
+
+  The secretary also creates a PR with a proposal saying (roughly) 
+   
+    GHC202x contains the following extensions in addition to those in GHC202(x-1):
+    
+    * (none yet)
+    
+    and removes these extensions
+   
+    * (none yet)
+
+* In order to gather data on the criterium “widespread usage”, the secretary creates a tally of which extensions are used how often on hackage.
+
+* In order to gather data on the criterium “community support”, the secretary runs a public poll on a suitable platform for one week where anyone
+  can vote in favor or against the inclusion of a given extension.  
```

Comment:
It's not exactly the same, but I have asked "Which language extensions would you like to be enabled by default?" in the state of Haskell survey for the past three years:

- https://taylor.fausak.me/2019/11/16/haskell-survey-results/#s2q5
- https://taylor.fausak.me/2018/11/18/2018-state-of-haskell-survey-results/#question-038
- https://taylor.fausak.me/2017/11/15/2017-state-of-haskell-survey-results/#question-21

I plan to ask it again this year. If there's something I can do to make it more useful to this GHC20xx process, please let me know!

Reply:
@simonmar actually just mentioned that, and maybe that is simply enough to get some community data. We could make the process flexible in this point:
```suggestion
* In order to gather data on the criterium “community support”, the secretary runs a public poll on a suitable platform for one week where anyone
  can vote in favor or against the inclusion of a given extension, or points the committee to a suitable existing survey result.
```

Reply:
If I’d do a dedicated poll, I might ask for “Yes please”/“Don’t care”/“No, please don’t” to also gauge how much resistance it might have. But maybe not worth the bother.

Reply:
I like how that sounds! Other people have asked for something like this before. I've held off because it's already an intimidating question. But I think I could create a simple enough UI/UX if "Yes please" is 👍/upvote, "No, please don't" is 👎/downvote, and "Don't care" is no vote at all.